### ISSUE TITLE
Please consider retrying ExpiredToken

### ISSUE THREAD

**Issue by grrtrr** [awslabs/aws-c-s3#471]
### Describe the feature

Add `ExpiredToken` to the list of retryable errors.

### Use Case

When processing is slow, it is possible to overrun the time budget of an STS token (usually 1 hour).

The STS tokens are refreshed only when a new request is made, or a request is retried.
This is by virtue of calling the delegated STS token provider, which does a `RefreshIfExpired`.

_Under healthy conditions, retrying a request with an expired refreshable token has a likelihood to succeed on retry_.
In the worst case there are a number of useless retries, which seems worth the expense.

### Proposed Solution

We have seen `ExpiredToken` under conditions similar to `RequestTimeout` (#457) and so propose adding `ExpiredToken` to the list of retryable exceptions.

### Why this would work

The STS tokens are refreshed each time `.sign_request = aws_s3_meta_request_sign_request_default` is called, by virtue of `aws_sign_request_aws`, which calls `aws_credentials_provider_get_credentials` on the (STS) credential provider delegated from the C++ SDK to the C libraries.

- On retry, `s_s3_client_retry_ready` is called (but only if `AWS_S3_CONNECTION_FINISH_CODE_RETRY` is set).
- This calls `aws_s3_meta_request_prepare_request`, which calls `s_s3_meta_request_schedule_prepare_request_default`,
- which schedules `s_s3_meta_request_prepare_request_task`, which calls `s_s3_meta_request_on_request_prepared`,
- which calls `s_s3_meta_request_sign_request`, which calls `aws_s3_meta_request_sign_request_default`,
- which calls `aws_sign_request_aws`,
- which then _refreshes the credentials_ (part of what the STS token provider does when `GetCredentials()` is called).

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by grrtrr** [comment]
### May require different document parser

The exception for the `ExpiredToken` seems to follow different conventions. I enabled `ExpiredToken` to be retried. The requests still fail with
```console
[ERROR] 2024-12-03 19:11:46.775 S3MetaRequest [140546567829056] id=0x7fd33202d280 Meta request failed from error 1051 (socket is closed.). (request=0x7fd38f227000, response status=206). Try to setup a retry.
[ERROR] 2024-12-03 19:11:46.848 S3MetaRequest [140546521687616] id=0x7f54bde52600 Meta request cannot recover from error 14343 (Invalid response status from request). (request=0x7fd38f229d00, response status=400)
[ERROR] 2024-12-03 19:11:46.896 S3MetaRequest [140546326656576] id=0x7f54baeb7280 Meta request cannot recover from error 14343 (Invalid response status from request). (request=0x7fd38f218500, response status=400)
I20241203 19:11:46.948966 140547094928704 shuffle.cc:113] Memory usage: 438.854GB,  elements in shuffle buffer: 5622625
terminate called after throwing an instance of 'av::CheckException'
  what():  Check failure at perception/dataset/tensor_group_io.cc:145:
  Expected: 'x is ok', with x := 'blobstore::write_blob(outfile, all_data)' [av::status::Status]
  x = PutObject() failed
  where: cloud/aws/s3/s3_streambuf.cc:100
  extra: s3://aurora-cloud-swe-prod-batch-artifacts/opt/56c8c46d/logs/42456e37-44a6-4b43-a6d3-ff5db4062cbf/7a52f31082caa1e77975e5d086f2465c.rats: HTTP response code: 400
Resolved remote host IP address:
Request ID: VKNKMR8D6BES88Q8
Exception name: ExpiredToken
Error message: Unable to parse ExceptionName: ExpiredToken Message: The provided token has expired.
7 response headers:
connection : close
content-type : application/xml
date : Tue, 03 Dec 2024 19:11:46 GMT
server : AmazonS3
transfer-encoding : chunked
x-amz-id-2 : PO/hiz8R75/RxqEV59lf92LRWAseXIFX8t3vteqzS447lB6ihhQOGIQykOQM85Dy/FWkKoY5Auw=
x-amz-request-id : VKNKMR8D6BES88Q8

**Comment by grrtrr** [comment]
### Tested to fix the problem with an overloaded S3 server

We added retries for `ExpiredToken` as suggested above and tested with a bucket where the endpoints regularly close connections, return `500`, causing stalling, which produces the problem.
The job started at 1:17 and stalled for over 1 hour. Hence at 2:17, the STS token expired.

We observed below that retrying on `ExpiredToken` allowed the job to recover.
```console
[2024-12-06 01:17:26,325][root][INFO] Invoking command:  <redacted> # start of job
[ERROR] 2024-12-06 01:23:41.999 S3MetaRequest [140592887621184] id=0x7fddcc63a400 Request failed from error 2058 (The connection has closed or is closing.). (request=0x7fde41cb1c80, response status=0). Try to setup a retry.
[ERROR] 2024-12-06 01:23:50.358 S3Endpoint [140592898111040] id=0x7fde59b5c150: Could not acquire connection due to error code 2058 (The connection has closed or is closing.)
[ERROR] 2024-12-06 01:30:08.934 S3MetaRequest [140592659035712] id=0x7fddca63a800 Request failed from error 2058 (The connection has closed or is closing.). (request=0x7fddc8610500, response status=0). Try to setup a retry.
[WARN] 2024-12-06 01:36:26.418 S3Client [140592843585088] id=0x7fde59b8a380 Client upload part timeout rate is larger than expected, current timeout is 1127, bump it up. Request original timeout is: 941
#
# Many similar such errors - retries, closed connections, and several 500 "Internal Server Error" cases during the next hour ...
#
[ERROR] 2024-12-06 02:17:31.980 S3MetaRequest [140592707257920] id=0x7f602f0a0000 Request failed from error 14370 (Token expired (needs a refresh).). (request=0x7fde41c76c80, response status=400). Try to setup a retry.
[ERROR] 2024-12-06 02:17:32.064 S3MetaRequest [140592614995520] id=0x7f602f0a0000 Request failed from error 14370 (Token expired (needs a refresh).). (request=0x7fde41c70f00, response status=400). Try to setup a retry.
```
Job then completed successfully.

**Comment by DmitriyMusatkin** [comment]
Do you know whats causing the stalling? Im not quite sure I follow what causes the creds to become expired. 
aws-c-s3 should signing the request last thing before it puts the request on the wire. every time request is signed, it first checks how long the creds are valid for and if its less than 5 min, it will block signing until it refreshes creds. 
We did recently (as of 4 months ago) bump that expiration bound check from 10s to 5min, since 10s was too small. Do you have the version number of crt on which you are seeing the problem? I think overall its fine to retry here, but want to make sure retry is not masking some bigger problem.

**Comment by grrtrr** [comment]
> Do you know whats causing the stalling?

This is a condition that several SDKs suffer from. I had put a list of these into #464 before realizing that #457 solved a related issue - retry on `RequestTimeout`.
`RequestTimeout` happens frequently under such conditions, due to the S3 stalling.
The stalling has been a long-term problem in 2 of our buckets. Sometimes tasks that finish in under 20 minutes drag along over 3 hours.

I had initially blamed the outdated state of our support libraries, but the problem persisted after upgrading `aws-crt-cpp` to `v0.29.3`, using all the related dependencies that this version uses (`aws-c-s3` at `v0.7.0`).

**Comment by graebm** [comment]
> ...aws-c-s3 should signing the request last thing before it puts the request on the wire...

Signing is NOT the last thing. Currently, the [stages](https://github.com/awslabs/aws-c-s3/blob/2e2a20cc2ed36a04e968bd1d9b40f2683e3c9e94/source/s3_client.c#L1702-L1705) of each HTTP request are:
- [Preparing](https://github.com/awslabs/aws-c-s3/blob/2e2a20cc2ed36a04e968bd1d9b40f2683e3c9e94/include/aws/s3/private/s3_client_impl.h#L419-L421) (signing happens at the [end of this stage](https://github.com/awslabs/aws-c-s3/blob/2e2a20cc2ed36a04e968bd1d9b40f2683e3c9e94/source/s3_meta_request.c#L716-L731))
- [Queued](https://github.com/awslabs/aws-c-s3/blob/2e2a20cc2ed36a04e968bd1d9b40f2683e3c9e94/include/aws/s3/private/s3_client_impl.h#L410-L412) (waiting for slot in the HTTP connection pool)
- Networking (Actually using HTTP connection)
- Streaming-waiting (Done, but out of order so can't be delivered to user yet)
- Streaming-response (Done, in order, being delivered to user)

So if it's a slow machine, the requests can sit in the Queued stage for a long time, already signed, waiting for a slot in the HTTP connection pool.

Garret's right, this probably IS a problem in any SDK where signing happens before waiting for a slot in a finite-sized HTTP connection pool.

If we delayed signing until the last possible moment that it needs to be submitted to the HTTP connection, we could reduce these timeout issues. But it's a non-trivial change, and might mess with our throughput on fast machines...

**Comment by waahm7** [comment]
Thank you for the PR. It has been released in https://github.com/awslabs/aws-c-s3/releases/tag/v0.7.7. We have also backlogged the task to move signing closer to the actual HTTP request to mitigate this issue further.

**Comment by grrtrr** [comment]
### Why retrying `ExpiredToken` is not a good idea

While #472 solved the problem of stale STS tokens for us, it has issues with non-retryable credential providers, such as `EnvironmentAWSCredentialsProvider` or `ProfileConfigFileAWSCredentialsProvider`:  
- retrying is futile since these providers do not automatically refresh credentials.
- The result is that `PutObject/GetObject/CopyObject` _hang indefinitely_ once credentials expire.
- The server response (OK or `ExpiredToken`) is needed to determine whether these credentials are still valid (so can not know a priori).

As per comments from @graebm and @waahm7, the _best way forward_ would be to
1. Move signing to the last possible moment to avoid the situation that credentials of queued requests become stale.
2. Revert #472 - as it is in hindsight not a good solution (solving one problem by introducing another one).

Other approaches look more complex - there are no  API methods to determine whether a given credential provider supports automatic refresh (which would justify retrying on `ExpiredToken`).

Until this is resolved, we intend to _work around_ this by checking for STS provider [hints](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) (`AWS_ROLE_ARN`, `AWS_WEB_IDENTITY_TOKEN_FILE`). But that also is _brittle_, since the STS credential provider can also be set up via `role_arn / web_identity_token_file` in the configuration file.

**Comment by TingDaoK** [comment]
I think the problem is not really an S3 client problem.
If the credentials provided is about to expire, no matter how close we move the signing to the http, it still may expire before server side receives it.

The proper fix in my opinion for this is apply the proactive refresh for the token to make sure the token provided have at least couple mins.

**Comment by grrtrr** [comment]
@TingDaoK - thank you for looking into this. The problem happens for _all_ credential providers that do not automatically refresh (e.g. environment credentials provider, or `~/.aws/credentials`).

Please see @graebm 's comment [above](https://github.com/awslabs/aws-c-s3/issues/471#issuecomment-2528840031); the _key problem_ is:
- signing happens _too early_ in `aws-c-s3`, so that
- if processing is slow and requests stay in the queue for _too long_, the credential token is _expired_ when the request is finally placed onto the wire.

https://github.com/awslabs/aws-c-s3/pull/472 was added to work around this key problem, but causes _hanging_ on non-refreshing credential providers,
- it should be reverted anyway as soon as signing happens later,
- _it would not harm to revert it now_ (avoids the problem that `Get/Put/CopyObject` hang when credentials from a non-refreshing credential provider expire).

**Comment by TingDaoK** [comment]
I understand what you said. I was to point out two things:
1. Without reverting the retry on expired token, it should NOT cause `Get/Put/CopyObject` to hang when credentials from a non-refreshing credential provider expire. If it leads to the hang, please provide us more information to dig into it.
2. The **key problem** to your issue is **NOT** the signing happens to early in `aws-c-s3`. It is the provided credentials provider should proactive refresh the credentials.

The reason for #1 is that I cannot reproduce the problem, it failed successfully for me after the retry finishes.

The reason for #2 is:
- moving signing closer to http can **help but NOT** gonna solve the issue in the end. Since in your credentials provider, you only refresh when the credentials expired, you can still meet the case where the credentials is like ms away from expiration, even if we send the request out right after we sign the request, the token can still expire before it reaches the server. Like @graebm said, _we could reduce these timeout issues_, but it won't rule it out.
- the time waiting in the queue is the network time for the request, as long as we follow the SDKs credentials provider's approach to refresh the credentials 5mins before the token expires, the network time is negligible.

**Comment by grrtrr** [comment]
@TingDaoK - it seems there is a misunderstanding here.

1. _To reproduce the problem_: please use `~/.aws/credentials` with expired credentials. Hangs in 100% of cases.
2. Can you please check with @graebm, the possibility for timeout should be reduced since this is a lower-level library on which other services depend.

We did see the problem of expired tokens of requests with STS token provider - _this does have automatic refresh_, still the request failed with `ExpiredToken`  for a _small percentage_ of requests. Hence asking whether we can improve on the situation in (2).

The time spans are not necessarily milliseconds. Under very high load this can be _in the order of multiple minutes up to 1 hour_. 
We have seen this happening when networking and processing load was very high.

**Comment by TingDaoK** [comment]
Let's focus on the first issue here.

I did, I used an expired credentials as the environment variable, which won't be able to refresh. And it's not hanging.

```
[TRACE] [2025-05-05T20:32:33Z] [0000000170443000] [S3MetaRequest] - response body: 
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>ExpiredToken</Code><Message>The provided token has expired.</Message>

xxx

[WARN] [2025-05-05T20:32:33Z] [0000000170443000] [exp-backoff-strategy] - id=0x600003ed0380: token 0x145905cf0 has exhausted allowed retries. Retry count 5 max retries 5
[ERROR] [2025-05-05T20:32:33Z] [0000000170443000] [standard-retry-strategy] - token_id=0x6000036c9090: error occurred while scheduling retry: aws-c-io: AWS_IO_MAX_RETRIES_EXCEEDED, Retry cannot be attempted because the maximum number of retries has been exceeded..
[ERROR] [2025-05-05T20:32:33Z] [0000000170443000] [S3Client] - id=0x145c05890 Client could not retry request 0x145b054b0 for meta request 0x124e063d0 with token 0x6000036c9090 due to error 1061 (Retry cannot be attempted because the maximum number of retries has been exceeded.)
```

So, it retries 5 times and in the end, it fails.

**Comment by TingDaoK** [comment]
The second issue

> We did see the problem of expired tokens of requests with STS token provider

Are you using the the STS token provider from SDKs or your own delegate one?

In your original issue description, you mentioned
>  calling the delegated STS token provider, which does a RefreshIfExpired.

That's what I want to point out.

**Comment by grrtrr** [comment]
> Let's focus on the first issue here.
> 
> I did, I used an expired credentials as the environment variable, which won't be able to refresh. And it's not hanging.
> 
> 
> So, it retries 5 times and in the end, it fails.

Apologies, I should have added that not `aws-c-s3` per se is hanging, but rather `aws-cpp-sdk` with the `S3CrtClient`. 
The `aws-c-s3` test cases are likely ok, but the ones for the `aws-cpp-sdk` are not; the calls do hang.

**Comment by TingDaoK** [comment]
We will check with CPP SDK about the hanging issue. Thanks for the information.


For the second issue,

Moving the signing closer to http is basically the same as the credentials provider refresh the credentials that is about to expire. 
Look into the code, the request after signing will stay in the queue waiting for the previous request that sent to HTTP level.
So, the time it spent in the queue will be basically the time a request spent on HTTP, which based on the part size and the network throughput, but I won't think it will take more than one minutes. 

Looks like each SDKs have its own behavior for the credentials provider to refresh, specifically:
1. The STS credentials provider in C++ SDK only refresh 5 secs before expires. https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/auth/STSCredentialsProvider.cpp
2. Java SDK v2 block refresh 1 mins before expire, will async refresh 5 mins before expire https://github.com/aws/aws-sdk-java-v2/blob/master/services/sts/src/main/java/software/amazon/awssdk/services/sts/auth/StsCredentialsProvider.java
4. CRT 5 mins https://github.com/awslabs/aws-c-auth/pull/247
5. botocore, 10 mins. https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L64-L65

Overall, I think have the credentials provider to refresh earlier is a cheaper and safer approach, while moving signing could impact performance, as the HTTP level will probably need to wait for the signing when a connection comes to be available, while it doesn't need to do currently.

Also, given the SDKs has different behavior for credentials provider, it's reasonable for us to retry on the expired token.

To summarize:
My take is keep the retry on expired token, focus on solving the hanging issue for C++ SDK.

**Comment by grrtrr** [comment]
@TingDaoK - thank you for providing these helpful details.

The _second issue_ is now clear:
- We were using the `STSAssumeRoleWebIdentityCredentialsProvider` with the standard 5 second grace period.
- Under load, the _queue time_ was perhaps 1 minute.
- A _fraction_ of the requests failed with `ExpiredToken` - the 5 second grace period expired while the request was queued.

For the _first issue_, I had left out an _important detail_: 
- our _retry strategy_ is set to retry for up to 30 minutes (this is done in order to deal better with prefix splitting).
- Hence the "hanging" problem is not caused by the AWS C++ SDK, but by the retry strategy.

The _root cause_ for the first issue is that requests are retried. The following 2 credential providers would seem to hang, but in reality just take a long time until all retries are exhausted:
- `EnvironmentAWSCredentialsProvider` with stale credentials,
- `ProfileConfigFileAWSCredentialsProvider` with stale credentials.

#### Proposal

1. _For the first issue_: 
   Since we can not assume a short number of retries (for the default case, I agree - 5 retries do not really matter), it would be better to _revert #472.
2. _For the second issue_: 
   Bring the AWS C++ CredentialsProvider on par with `aws-c-auth`, to also use a grace period of 5 minutes (rather than 5 seconds).

Would you be ok with that?

**Comment by grrtrr** [comment]
Thank you @waahm7 - put up https://github.com/aws/aws-sdk-cpp/pull/3415 for the second one (which mimics https://github.com/awslabs/aws-c-auth/pull/247).

**Comment by sidewinder12s** [comment]
@TingDaoK Would it be possible to surface to the broader SDK team if they can unify/standardize on this behavior across SDKs? 

These types of issues with credentials and the provider chains have been really insidious for any customers who use more than one of these SDKs in AWS (And has been the case for our organization since we migrated to AWS)

**Comment by TingDaoK** [comment]
@sidewinder12s Yeah, I agree this inconsistent behavior across SDKs will cause a lot confusion. And it was also a surprise to me when I dig into C++ SDK and found it out.

I'll bring this up to the broader SDK team, thanks for the feedback!

**Comment by waahm7** [comment]
I have released https://github.com/awslabs/aws-c-s3/releases/tag/v0.7.18 which reverts the retry token PR.
